src/common/sts.c

```diff
@@ -0,0 +1,151 @@
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+#include <fcntl.h>
+#include <time.h>
+
+#include "cfgfiles.h"
+#include "sts.h"
+#include "util.h"
+
+#define STS_CONFIG_FILE "sts.conf"
+
+GSList *profiles = NULL;
+
+struct sts_profile *
+sts_find (const char* host)
+{
+	time_t now;
+	GSList *next;
+	struct sts_profile *nextprofile;
+
+	now = time (NULL);
+	for (next = profiles; next; next = next->next)
+	{
+		nextprofile = (struct sts_profile *)next->data;
+		if (now >= nextprofile->expiry)
+			continue; /* Profile has expired. */
+
+		if (!g_strcmp0 (host, nextprofile->host))
+			return nextprofile; /* We found the right profile! */
+	}
+
+	/* No profile for this host. */
+	return NULL;
+}
+
+void
+sts_load (void)
+{
+	char buf[256];
+	int fh;
+	struct sts_profile *profile;
+
+	fh = hexchat_open_file (STS_CONFIG_FILE, O_RDONLY, 0, 0);
+	if (fh < 0)
+		return; /* Filesystem not readable. */
+
+	while (waitline (fh, buf, sizeof buf, FALSE) != -1)
+	{
+		profile = sts_new ();
+		if (sscanf (buf, "%s %u %ld", profile->host, &profile->port, &profile->expiry) != 3)
+		{
+			/* Malformed profile; drop it. */
+			g_free (profile);
+			continue;
+		}
+
+		sts_store (profile);
+	}
+}
+
+struct sts_profile *
+sts_new (void)
+{
+	struct sts_profile *profile;
+
+	profile = malloc (sizeof (struct sts_profile));
+	profile->host[0] = 0;
+	profile->port = 0;
+	profile->expiry = 0;
+	return profile;
+}
+
+GHashTable *
+sts_parse_cap (const char* cap)
+{
+	char **entries, **currentry;
+	char *value;
+	GHashTable *table;
+
+	table = g_hash_table_new (g_str_hash, g_str_equal);
+	entries = g_strsplit (cap, ",", 0);
+	for (currentry = entries; *currentry; ++currentry)
+	{
+		value = strchr (*currentry, '=');
+		if (value)
+			g_hash_table_insert (table, g_strndup (*currentry, value - *currentry), g_strdup (value + 1));
+	}
+
+	g_free (entries);
+	return table;
+}
+
+void
+sts_save (void)
+{
+	char buf[512];
+	int fh;
+	const GSList *next;
+	struct sts_profile *nextprofile;
+	time_t now;
+
+	fh = hexchat_open_file (STS_CONFIG_FILE, O_TRUNC | O_WRONLY | O_CREAT, 0600, XOF_DOMODE);
+	if (fh < 0)
+		return; /* Filesystem not writable. */
+
+	now = time (NULL);
+	for (next = profiles; next; next = next->next)
+	{
+		nextprofile = (struct sts_profile *)next->data;
+		if (now >= nextprofile->expiry)
+			continue; /* Profile has expired. */
+
+		g_snprintf (buf, sizeof buf, "%s %u %ld\n", nextprofile->host, nextprofile->port,
+			(unsigned long)nextprofile->expiry);
+		write (fh, buf, strlen (buf));
+	}
+
+	close (fh);
+}
+
+void
+sts_store (struct sts_profile *profile)
+{
+	profiles = g_slist_append (profiles, profile);
+}
+
+void sts_update_expiry (const char *host, time_t newexpiry, struct sts_profile *incomplete_profile)
+{
+	struct sts_profile *profile;
+
+	profile = incomplete_profile ? incomplete_profile : sts_find (host);
+	if (profile)
+		profile->expiry = time (NULL) + newexpiry;
```

Comment:
This can be larger than `profile->expiry`, no validation is done.

Reply:
How would you validate this in C? In C++ you would check if `std::numeric_limits<time_t>::max()` minus `newexpiry` is greater than the return value of `time(NULL)` but there's no TIME_MAX equivalent in C. The glib checked add functions don't work for time_t either.